### ISSUE TITLE
ouput not correct

### ISSUE THREAD

**Issue by glepnir** [BrentFarris/LuaAsyncAwait#1]
@BrentFarris  First thanks for a min example of async/await in lua .but I think it's not async correctly and you missed add `update_frame ` into while loop. my test code

```lua
local frameCounter = 0
--- Imagine this funciton is called every frame in a game engine
function update_frame()
	frameCounter = frameCounter + 1
end

-- Imagine this function is called once when the game starts
function init_game()
	print("Starting async")
	async(do_async, 6)
	print("Async was started, I'm not waiting for it")
end

-- The async function to be called from anywhere in the game
function do_async(await, countTo)
	print("Waiting 1000 frames")
	-- The code stops here and waits for async.resolve or async.reject before continuing
	local msg = await(wait_for_frames, 10)
	print(msg)
	for i=1, countTo do
		print("Printing number", i)
	end
	print("Async work complete")
end

-- The function to be awaited on, notice `await` is the first argument
function wait_for_frames(await, frames)
	-- Basically wait 1000 frames in the game
	while frameCounter < frames do
    update_frame()
		--Nothing
	end
	-- This will continue to the caller and return the argument as the result
	await.resolve("All done here")
end

init_game()
```
## Output

```
Starting async
Waiting 1000 frames
All done here
Printing number 1
Printing number 2
Printing number 3
Printing number 4
Printing number 5
Printing number 6
Async work complete
Async was started, I'm not waiting for it
```
See the ouput. I found it still sync ?

**Comment by BrentFarris** [comment]
You are correct, the example given is not a good example. There shouldn't be an infinite loop in the async call because it will halt the thread. It is hard to show a reasonable example. Imagine you were to start an async call that waits for a button to be clicked, then it continues. Or you start an async call and wait for a response from the network, these would be better examples of async/await. I think I'll put together an example of a fake network request to show how to use this correctly (without infinite loops).

**Comment by BrentFarris** [comment]
I've updated the example code to be a fake network request example:

```lua
function test_async()
        print("Starting async")
        async(do_async, "https://github.com/BrentFarris/LuaAsyncAwait")
        print("Async was started, I'm not waiting for it")
end

local fakeNet = nil
-- The async function to be called from anywhere in the game
function do_async(await, addr)
        print("Doing a fake network request to: "..addr)
        local req = {
                done = function(res) end,
                sim = coroutine.create(function()
                        local clock = os.clock
                        local t0 = clock()
                        local fakeDelay = t0 + 3.0 -- Seconds (slow network :P)
                        while clock() - t0 < fakeDelay do
                                coroutine.yield()
                        end
                        -- Got response from network
                        fakeNet.done("Hello from the fake internet")
                end)
        }
        fakeNet = req
        -- The code stops here and waits for async.resolve or async.reject before continuing
        local msg = await(wait_for_net, req)
        print("Message from fake network: "..msg)
        print("Async work complete")
end

-- The function to be awaited on, notice `await` is the first argument
function wait_for_net(await, req)
        req.done = function(res) await.resolve(res) end
        print("Going to wait on the network to finish...")
end

test_async()

local clock = os.clock
local d0 = clock()
local delay = d0 + 5.0
while clock() - d0 < delay do
        -- Just keep the program alive
        coroutine.resume(fakeNet.sim)
end

--[[ OUTPUT
	Starting async
	Doing a fake network request to: https://github.com/BrentFarris/LuaAsyncAwait
	Going to wait on the network to finish...
	Async was started, I'm not waiting for it
-- 3 seconds pass
	Message from fake network: Hello from the fake internet
	Async work complete
]]
```

**Comment by glepnir** [comment]
looks like great. add status will be better :)

**Comment by glepnir** [comment]
@BrentFarris  Hi there.  in example `re.sim` filed can i pass a function into it? not coroutine?
